wdl-lint/tests/lints/command-mixed-indentation/source.wdl

```diff
@@ -0,0 +1,39 @@
+## This is a test of having mixed indentation in command sections.
+
+version 1.1
+
+task test1 {
+    command <<<
+        this line is ~{
+		    if true
+		    then "split across multiple lines with mixed indentation"
+		    else "by a placeholder"
+	    } but is all one literal line in the command text
+        this line has a continuation \
+ 		   and should not be a warning
+        this line is prefixed with ~{"spaces and has tailing mixed indentation"}  		
+		this line is prefixed with ~{"tabs"}
+      	this line is prefixed with mixed indentation
+        this line is prefixed with spaces and has trailing mixed indentation  		
+			   >>>
+
+    runtime {}
+}
+
+task test2 {
+    command {
+        this line is ~{
+		    if true
+		    then "split across multiple lines with mixed indentation"
+		    else "by a placeholder"
+	    } but is all one literal line in the command text
+        this line has a continuation \
+ 		   and should not be a warning
```

Comment:
I think this _should_ be a warning. I'm taking the perspective of a user. If I write the following WDL:
```
command {
    my-tool \
    arg1 \
    arg2 \
    arg3
}
```
I would expect all those indents of size 4-spaces to be stripped from my command before execution. If I expect them to be stripped, it follows that they can't be mixed.

Reply:
My understanding is that line continuation stripping is different from command whitespace stripping, though. It should be stripping _all_ leading whitespace on the next line until the first non-whitespace character, so it doesn't matter if it's mixed or not.

Leading whitespace in a command line is stripped by _common_ whitespace, which would then necessitate you don't mix the whitespace characters.

Reply:
Does that come from the spec? What about the following WDL:
```
command {
    my-tool \
        arg1
}
```
I would (perhaps naively) assume it would execute the following command text:
```

my-tool \
    arg1
```

Reply:
The spec calls `\<newline>` a continuation, but that's it.

If it's anything like a line continuation in other programming languages, it should treat that as `my-tool arg1` with no newline or leading whitespace.

Reply:
Here's where I think WDL is different. I don't think the line continuation should be parsed by the WDL engine (or us). It should be passed along as part of the literal command text. If the spec doesn't say anything about how to parse line continuations, I think that's because they aren't meant to be parsed.

Check out this `command` file produced by `miniwdl`:
```
$ cat SJBALL034267_D1.WholeGenome/_LAST/call-multiqc/command

        set -euo pipefail

        # set ENV variables for `multiqc`
        export LC_ALL=C.UTF-8
        export LANG=C.UTF-8

        echo "/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.ValidateSamFile.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.flagstat.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.instrument.tsv
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.readlength.tsv
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.encoding.tsv
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.endedness.tsv
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome_fastqc.zip
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.CollectAlignmentSummaryMetrics.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.CollectInsertSizeMetrics.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.QualityScoreDistribution.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.kraken2.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.whole_genome.mosdepth.summary.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.whole_genome.mosdepth.global.dist.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.global_PHRED_scores.tsv
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.exon.mosdepth.summary.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.CDS.mosdepth.summary.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.UTR.mosdepth.summary.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.exon.mosdepth.region.dist.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.CDS.mosdepth.region.dist.txt
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/SJBALL034267_D1.WholeGenome.UTR.mosdepth.region.dist.txt" > file_list.txt

        # --strict is too strict. It causes errors due
        # to how our config adds 'custom-content' to the report.
        # Leaving this here as a warning not to try putting it back.
        # --require-logs might be useful at some point, but as of now,
        # it would cause errors. It could replace the check currently
        # run after multiqc is finished.
        # TODO: lots of other options to consider supporting.
        multiqc -v \
            --no-ansi \
            -c /mnt/miniwdl_task_container/work/_miniwdl_inputs/0/multiqc_config_hg38.yaml \
            --file-list file_list.txt \
            -o SJBALL034267_D1.WholeGenome.multiqc

        if [ ! -d SJBALL034267_D1.WholeGenome.multiqc ]; then
            >&2 echo "MultiQC didn't find any valid files!"
            exit 1
        fi

        tar -czf SJBALL034267_D1.WholeGenome.multiqc.tar.gz SJBALL034267_D1.WholeGenome.multiqc

```

Reply:
Which actually hasn't been properly whitespace stripped. Could be a problem with our code though.

Reply:
(PSA @adthrasher , it looks like miniwdl is not consistently stripping our whitespace as expected. I haven't identified the exact WDL constructions that cause it, but it's only a problem for some of our commands. Not a big issue as these commands aren't whitespace sensitive, but we should report to miniwdl just to be thorough)

Reply:
I see. It's really bizarre to me that it would call it out as a line continuation, though, if it didn't want engines to do something with it.

If they _hadn't_ called out `\<newline>` as a line continuation, it would have had the exact same effect: not an escape sequence and therefore a _literal backslash_ at the end of the line to be interpreted by _the shell_ as a line continuation.

At any rate, I'll chalk this up as vague warning in the spec (if it means it to be a line continuation _in the shell script_, it should probably say so) and fix the rule.

🤷